### PR TITLE
Elements additions

### PR DESCRIPTION
**Pull request by Panos512**
- Adds MalformedQuery element which is handled on various installations.
- Adds Wildcard element to support queries containing wildcard characters (`*` and `#`).

### CODE REVIEW

File: invenio_query_parser/ast.py [hunk 1]
```diff
@@ -1,7 +1,8 @@
 # -*- coding: utf-8 -*-
 #
 # This file is part of Invenio-Query-Parser.
-# Copyright (C) 2014, 2016 CERN.
+#
```
Comment (jirikuncar) [inline invenio_query_parser/ast.py]: not needed

**Comment by jirikuncar** [comment]
We should also check that Elastic query builder works with the new ast element.

**Comment by jirikuncar** [comment]
@Panos512 Please test new element with all walkers and contrib/elasticsearch. Thanks

**Comment by egabancho** [comment]
In principle I don't see any issue to include this, just a few questions

* which difference does it make? I meant, with the current implementation the query it generates seems the same ...
* why also `#` as wildcard and not only `*`?
* what about `*foo`?

and as @jirikuncar mentioned 
>  Please test new element with all walkers and contrib/elasticsearch

**Comment by jacquerie** [comment]
> * why also `#` as wildcard and not only `*`?

Legacy from SPIRES: the two symbols were equivalent there, and they are still equivalent in the current INSPIRE: https://inspirehep.net/search?ln=en&ln=en&p=find+j+%22Phys.Rev.Lett.%2C105*%22 and https://inspirehep.net/search?ln=en&ln=en&p=find+j+%22Phys.Rev.Lett.%2C105%23%22 and

**Comment by Panos512** [comment]
> * which difference does it make? I meant, with the current implementation the query it generates > seems the same ...

For our case we treat `WildcardQuery` differently (https://github.com/inspirehep/inspire-next/blob/master/inspirehep/modules/search/walkers/elasticsearch.py#L94), and we are going to use the `MalformedQuery` element to show messages to users when they perform malformed queries.

> * what about `*foo`?

The problem with accepting wildcards everywhere in a word is that there are some particles that contain wildcard characters (e.g. `D/sJ*(2632)`). So accepting them only in the end of a word is partly solving the problem.

and as @jirikuncar mentioned 
>  Please test new element with all walkers and contrib/elasticsearch

Apart from `invenio-query-parser` tests, that pass at the moment, do we test elasticsearch query generation somewhere else?

**Comment by egabancho** [comment]
> The problem with accepting wildcards everywhere in a word is that there are some particles that contain wildcard characters (e.g. D/sJ*(2632)). So accepting them only in the end of a word is partly solving the problem.

IMHO this type of strings should be escaped using quotes to avoid ambiguity, i.e.  `"D/sJ*(2632)"`.
We have more cases like this, for example titles containing `and` or `or`, and not because of that we stopped using this keywords in the middle of queries.

> Apart from invenio-query-parser tests, that pass at the moment, do we test elasticsearch query generation somewhere else?

Test are passing indeed but, you didn't add any test to the [walkers](https://github.com/inveniosoftware/invenio-query-parser/blob/master/tests/test_walkers.py) to check that your change is not breaking them, which is what we are asking for.
In fact you didn't modify the walkers at all, that's why I wondered before which difference it makes 😉

--

Maybe if this is so inspire/spires centric you should consider updating [contrib/spires](https://github.com/inveniosoftware/invenio-query-parser/tree/master/invenio_query_parser/contrib/spires) instead (just a thought).

**Comment by Panos512** [comment]
> IMHO this type of strings should be escaped using quotes to avoid ambiguity, i.e.  `"D/sJ*(2632)"`.
> We have more cases like this, for example titles containing `and` or `or`, and not because of that we stopped using this keywords in the middle of queries.

I'll take a look into this. Indeed it could be done like this and it's the optimal solution for us. But are we sure that the users are going to escape words with special characters? Maybe we could white-list all known particles and ignore special characters in them ( I could resurrect #39 for this ). 

> Test are passing indeed but, you didn't add any test to the [walkers](https://github.com/inveniosoftware/invenio-query-parser/blob/master/tests/test_walkers.py) to check that your change is not breaking them, which is what we are asking for.
> In fact you didn't modify the walkers at all, that's why I wondered before which difference it makes 😉

That's true, I will update the walkers + tests.

> Maybe if this is so inspire/spires centric you should consider updating [contrib/spires](https://github.com/inveniosoftware/invenio-query-parser/tree/master/invenio_query_parser/contrib/spires) instead (just a thought).

The thing is that we need this functionality in the `invenio-syntax` and not only on `spires-syntax`.

**Comment by kaplun** [comment]
@jirikuncar I think @Panos512 has fullfilled the requests, and thus this PR is in good state to be merged.

File: invenio_query_parser/contrib/elasticsearch/walkers/dsl.py [hunk 2]
```diff
@@ -80,6 +81,24 @@ def visit(self, node, op):
     def visit(self, node):
         return node.value
 
+    @visitor(WildcardQuery)
+    def visit(self, node):
+        def query(keyword):
```
Comment (jirikuncar) [inline invenio_query_parser/contrib/elasticsearch/walkers/dsl.py]: I don't see tests for this part.

File: invenio_query_parser/contrib/elasticsearch/walkers/dsl.py [hunk 3]
```diff
@@ -80,6 +81,24 @@ def visit(self, node, op):
...
+            fields = self.get_fields_for_keyword(keyword, mode='p')
+            if len(fields) > 1:
+                res = Q('bool', should=[
```
Comment (jirikuncar) [inline invenio_query_parser/contrib/elasticsearch/walkers/dsl.py]: `return Q('bool', should=[`

File: invenio_query_parser/contrib/elasticsearch/walkers/dsl.py [hunk 4]
```diff
@@ -80,6 +81,24 @@ def visit(self, node, op):
...
+                        Q('query_string',
+                          query=value,
+                          default_field=k,
+                          analyze_wildcard=True) for k in fields])
+            else:
```
Comment (jirikuncar) [inline invenio_query_parser/contrib/elasticsearch/walkers/dsl.py]:
```diff
- else
+ return Q('query_string',
```

File: invenio_query_parser/ast.py [hunk 5]
```diff
@@ -182,3 +186,7 @@ class RegexValue(Leaf):
 
 class EmptyQuery(Leaf):
     pass
+
+
+class MalformedQuery(Leaf):
```
Comment (kaplun) [inline invenio_query_parser/ast.py]: Ideally yes. This is then needed to implement: https://github.com/inspirehep/inspire-next/pull/1182

File: invenio_query_parser/contrib/elasticsearch/walkers/dsl.py (hunk 2 above)
Comment (kaplun) [inline invenio_query_parser/contrib/elasticsearch/walkers/dsl.py]: Panos has now left. Can you maybe guide us on concretely what type of tests we would need to write?

File: invenio_query_parser/ast.py (hunk 5 above)
Comment (jirikuncar) [inline invenio_query_parser/ast.py]: In that case I don't see any test for it.

**Comment by lnielsen** [comment]
Closing PR as it's been stale for a long time.

**Pull request closed without merging**